### PR TITLE
Remove ref for internal action calls

### PR DESCRIPTION
**Pull request by laurentsimon**
Remove the ref we use to call internal actions, Instead, checkout the builder repo and calls them locally.

closes https://github.com/slsa-framework/slsa-github-generator/issues/880

### CODE REVIEW

File: .github/actions/generate-builder/action.yml
Comment (laurentsimon) [inline .github/actions/generate-builder/action.yml:36]: I will update `@main` to `@hash` once the PR is merged.

**Comment by laurentsimon** [comment]
@ianlewis appreciate feedback and questions, since this seems to undo some of the secure-checkout code. I want to be sure I'm not mis-understanding something.

File: .github/actions/secure-project-checkout-go/action.yml
Comment (asraa) [inline .github/actions/secure-project-checkout-go/action.yml:26]: doesn't this need a secure builder dir checkout before use?
same below for node. or maybe I'm missing something

File: .github/actions/secure-project-checkout/action.yaml
Comment (asraa) [inline .github/actions/secure-project-checkout/action.yaml:5]: nit: defaults

File: .github/workflows/builder_go_slsa3.yml
Comment (asraa) [inline .github/workflows/builder_go_slsa3.yml:248]: my earlier comment: after the buidl procedure, do we trust `./__BUILDER_CHECKOUT_DIR__`?
https://github.com/slsa-framework/slsa-github-generator/issues/880#issuecomment-1282771342

File: .github/actions/secure-project-checkout-go/action.yml
Comment (asraa) [reply .github/actions/secure-project-checkout-go/action.yml:26]: maybe it propogates since it's an action and run in the same environment? if so, maybe this action should validate that BUILDER_CHECKOUT_DIR is populated?

Comment (laurentsimon) [reply .github/actions/secure-project-checkout-go/action.yml:26]: you're correct, it assumes the builder repo is checkout'ed. The checkout is done in the main re-usable workflow by calling `uses: slsa-framework/slsa-github-generator/.github/actions/secure-builder-checkout`.

The actions run in the same context as the job, so it's present.

What do you mean by "maybe this action should validate that BUILDER_CHECKOUT_DIR is populated"? The action will fail if the directory is not present, so the validation is implicit. Do you have something else in mind?

Btw, BUILDER_CHECKOUT_DIR is not an env variable, because `uses:` does not accept them.

Let me know if this clarifies or still confusing.

Comment (asraa) [reply .github/actions/secure-project-checkout-go/action.yml:26]: > The action will fail if the directory is not present, so the validation is implicit.

Only for a clear error message: maybe can you add a comment to the action.yml saying it requires that the repository is populated? Just in case it hits an error when creating a new builder

File: .github/workflows/builder_go_slsa3.yml
Comment (asraa) [reply .github/workflows/builder_go_slsa3.yml:248]: Got it, resolved in the issue!

**Review by asraa — APPROVED** [review]
lgtm, after I understand now the user risk and interaction here.

File: .github/actions/secure-builder-checkout/action.yaml
Comment (ianlewis) [inline .github/actions/secure-builder-checkout/action.yaml:5]: I feel like secure-builder-checkout could call `detect-workflow` itself since the `repository` and `ref` here are always the results of `detect-workflow`.

That way we won't even need the `repository` and `ref` inputs.

File: .github/actions/secure-builder-checkout/action.yaml
Comment (ianlewis) [inline .github/actions/secure-builder-checkout/action.yaml:9]: If we pass this directly to `actions/checkout` it doesn't need to be fully formed. It could just be a tag or branch name.

File: .github/actions/secure-project-checkout/action.yaml
Comment (ianlewis) [reply .github/actions/secure-project-checkout/action.yaml:5]:
```suggestion
  # the token is not available to actions by default, so we need to
```

File: .github/actions/secure-project-checkout/action.yaml
Comment (ianlewis) [inline .github/actions/secure-project-checkout/action.yaml:6]:
```suggestion
  # share it explicitly. The token is needed to checkout private repositories.
```

File: .github/actions/secure-project-checkout/action.yaml
Comment (ianlewis) [inline .github/actions/secure-project-checkout/action.yaml:37]: @laurentsimon Since you mentioned `actions/checkout` pulls a tar, I wondered if that tar could somehow be overwritten with a dirty checkout such that `git log -1 --format='%H'` returns the right SHA but the checkout itself it dirty and contains different code.

Maybe not necessary for this PR but we could maybe add a check for `git status` or `git diff` in the future.

**Review by ianlewis — APPROVED** [review]

**Comment by ianlewis** [comment]
> @ianlewis appreciate feedback and questions, since this seems to undo some of the secure-checkout code. I want to be sure I'm not mis-understanding something.

I think this is fine. I wasn't totally confident `secure-checkout` was solving the issue properly anyway.

File: .github/workflows/builder_go_slsa3.yml
Comment (ianlewis) [inline .github/workflows/builder_go_slsa3.yml:192]: This unfortunately won't pass pre-submits because `secure-builder-checkout` only exists in this PR and not at `main` currently. We can either temporarily remove the required check in order to merge, or we'll need to add the secure-builder-checkout to the repo first before using it in the top level reusable workflows.

**Comment by ianlewis** [comment]
Related #795

File: .github/actions/secure-builder-checkout/action.yaml
Comment (laurentsimon) [reply .github/actions/secure-builder-checkout/action.yaml:5]: discussed offline: until now we could have done that. But this PR requires the top re-usable workflow to checkout the builder code, so we need the repository / ref information to be available.

File: .github/actions/secure-builder-checkout/action.yaml
Comment (laurentsimon) [reply .github/actions/secure-builder-checkout/action.yaml:9]: Thanks. I updated the description.

**Pull request merged**
